unumpy/tests/test_numpy.py

```diff
@@ -542,6 +595,31 @@ def test_functional(backend, method, args, kwargs):
         ret.compute()
 
 
+@pytest.mark.parametrize(
+    "method, args",
+    [
+        (np.c_, ([1, 2, 3], [4, 5, 6])),
+        (np.r_, ([1, 2, 3], [4, 5, 6])),
+        (np.s_, (slice(2, None, 2))),
+    ],
+)
+def test_class_getitem(backend, method, args):
+    backend, types, = backend
+    try:
+        with ua.set_backend(backend, coerce=True):
+            ret = method[args]
+    except ua.BackendNotImplementedError:
+        if backend in FULLY_TESTED_BACKENDS and (backend, method) not in EXCEPTIONS:
+            raise pytest.xfail(
+                reason="The backend has no implementation for this class."
+            )
+
+    if method is np.s_:
+        assert isinstance(ret, slice)
+    else:
+        assert isinstance(ret, onp.ndarray)
```

Comment:
That's interesting, do these return `numpy.ndarray`s for other backends as well? If so, that seems to be a bug in the backends.

Reply:
These classes are not dispatching their arguments and since the backends don't have them (at least Dask and Sparse) unumpy overrides them to NumPy classes which in turn convert the arguments into `numpy.ndarray`s. I think that they should dispatch their arguments but either way the NumPy classes will be operating on `ndarrays` from other backends (it works in some cases). Other option could be to just return `NotImplemented`.

Reply:
I will debug this, see what's going wrong, but if what you say is correct, `NotImplemented` is the way to go.

Reply:
Maybe create a list of not implemented classes in each backend and inside `overridden_class` check if `self` is in that list, if yes then return `NotImplemented`. I tried dispatching the list-like arguments in `__getitem__`. The Dask tests pass because of this behaviour:

```python
In [4]: a = da.from_array(onp.asarray([1, 2, 3]))                               

In [5]: res = onp.c_[a]                                                         
/Users/joaoferreira/miniconda3/envs/unumpy/lib/python3.7/site-packages/dask/array/core.py:1356: FutureWarning: The `numpy.ndim` function is not implemented by Dask array. You may want to use the da.map_blocks function or something similar to silence this warning. Your code may stop working in a future release.
  FutureWarning,

In [6]: res                                                                     
Out[6]: 
array([[1],
       [2],
       [3]])

In [7]: type(res)                                                               
Out[7]: numpy.ndarray
```

And Sparse tests fail because of this:

```python
In [8]: a = sparse.as_coo(onp.asarray([1, 2, 3]))                               

In [9]: res = onp.c_[a]                                                         
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-f3f273ef9019> in <module>
----> 1 res = onp.c_[a]

~/miniconda3/envs/unumpy/lib/python3.7/site-packages/numpy/lib/index_tricks.py in __getitem__(self, key)
    385             else:
    386                 item_ndim = ndim(item)
--> 387                 newobj = array(item, copy=False, subok=True, ndmin=ndmin)
    388                 if trans1d != -1 and item_ndim < ndmin:
    389                     k2 = ndmin - item_ndim

~/miniconda3/envs/unumpy/lib/python3.7/site-packages/sparse/_sparse_array.py in __array__(self, **kwargs)
    221         if not AUTO_DENSIFY:
    222             raise RuntimeError(
--> 223                 "Cannot convert a sparse array to dense automatically. "
    224                 "To manually densify, use the todense method."
    225             )

RuntimeError: Cannot convert a sparse array to dense automatically. To manually densify, use the todense method.
```

Even if they all passed I'm guessing that we want to avoid calling NumPy's methods/classes with `ndarray`s from other backends.

Reply:
This is fixed by the latest commits.